src/python/pants/backend/python/rules/python_create_binary.py

```diff
@@ -2,47 +2,77 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 from dataclasses import dataclass
-from typing import Optional
+from typing import Optional, Tuple
 
 from pants.backend.python.rules.pex import Pex
 from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
-from pants.backend.python.rules.targets import EntryPoint, PythonBinarySources
+from pants.backend.python.rules.targets import (
+    PexAlwaysWriteCache,
+    PexEmitWarnings,
+    PexIgnoreErrors,
+    PexIndices,
+    PexInheritPath,
+    PexRepositories,
+    PexShebang,
+    PexZipSafe,
+    PythonBinarySources,
+    PythonEntryPoint,
+    PythonPlatforms,
+)
 from pants.backend.python.targets.python_binary import PythonBinary
-from pants.build_graph.address import Address
 from pants.engine.addressable import Addresses
 from pants.engine.rules import UnionRule, rule
 from pants.engine.selectors import Get
-from pants.engine.target import Target
-from pants.rules.core.binary import BinaryImplementation, CreatedBinary
+from pants.rules.core.binary import BinaryConfiguration, CreatedBinary
 from pants.rules.core.determine_source_files import AllSourceFilesRequest, SourceFiles
 
 
 @dataclass(frozen=True)
-class PythonBinaryImplementation(BinaryImplementation):
-    required_fields = (EntryPoint, PythonBinarySources)
+class PythonBinaryConfiguration(BinaryConfiguration):
+    required_fields = (PythonEntryPoint, PythonBinarySources)
```

Comment:
The weird part of this is that there are two notions of Optional:

1) The target type doesn't have the Field registered to begin with.
2) The field is registered, but the user didn't explicitly define it.

This construct `required_fields`, paired with `tgt.get()`, deals with situation #1. 

--

If we go with `Optional[PexAlwaysWriteCache]`, then the call sites now must deal with both distinct notions of optionality, like this:

```python
if self.always_write_cache and self.always_write_cache.value is True:
    args.append("--always-write-cache")
if self.emit_warnings and self.emit_warnings.value is False:
    args.append("--no-emit-warnings")
```

Using `tgt.get()` is meant to avoid this clunkiness by saying "if your target type doesn't have this field in the first place, the default constructor for that `Field` will do just fine."

Reply:
OK, that makes sense. A tradeoff of oddness of requiredness marking for less verbose value checking. The values would also not be awkward to check if they were just newtypes. If there was no Field baseclass requirement, you could have:
```python
class Timeout(int):
  pass
```
That would satisfy the engine's need for distinct types if the field were ever a rule or Get parameter directly and it would kill te need to check for the existince of the wrapper before accessing its `value`. Now the newtype is the value, no wrapping. Could Field just be a required mixin for fields, ie:
```python
class Timeout(int, Field):
  alias = "timeout"
  default = 0
```

Just spitballing not meant to slow down this PR. There is awkwardness here and I think its saying something, but perhaps it just has to be lived with.

Reply:
> If there was no Field baseclass requirement, you could have:
>
> class Timeout(int, Field):
>   alias = "timeout"
>   default = 0

Interesting. Where the idea of subclassing a type gets weird is how you express complex types like this https://github.com/pantsbuild/pants/pull/9447/files#diff-ec9ce69cef47321746e1e039429a4b5dR126 (lines 126-163).

>  tradeoff of oddness of requiredness marking for less verbose value checking. 

Precisely. It is weird.

Reply:
You don't have to subclass a type, you just do it for the actual primitive types so they are not awkward to deal with. The case you point to does not look primitive, it is complex. It should be harder to deal with and wrap a value. If the call site using it is a bit awkward that just comports with the reality of that particular field type.

Reply:
> The case you point to does not look primitive, it is complex. 

Indeed. To clarify, `PrimitiveField` is distinguished from `AsyncField`. `PrimitiveField` only means "we're dealing with primitive Python values and don't need to use the engine for hydration."